src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -5,45 +5,121 @@
+                                     envpath: Puppet[:environmentpath],
+                                     facts: compile_options['facts']['values']) do |pal|
+            # TODO: Given we hide this from plan authors this current iteration has only
+            # the "required" data for now. Once we can get https://github.com/puppetlabs/bolt/pull/1770
+            # merged and promoted we can just use an empty hash.
+            fake_config = {
+              'transport' => 'redacted',
+              'transports' => {
+                'redacted' => 'redacted'
+              }
+            }
+            bolt_inv = Bolt::ApplyInventory.new(fake_config)
+            Puppet.override(bolt_inventory: bolt_inv) do
+              Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+              # This compiler has been configured with a node containing
+              # the requested environment, facts, and variables, and is used
+              # to compile a catalog in that context from the supplied AST.
+              pal.with_catalog_compiler do |compiler|
+                # TODO: PUP-10476 Explore setting these as default in PAL. They are the defaults in Puppet
+                Puppet[:strict] = :warning
+                Puppet[:strict_variables] = false
+                vars = Puppet::Pops::Serialization::FromDataConverter.convert(compile_options['variables']['values'])
+
+                compile_options['facts']['values'].keys.each {|fact_name| vars.delete(fact_name)}
+                # TODO: PUP-10476 Refactor PAL api such that we do not need to use private methods
+                pal.send(:add_variables, compiler.send(:topscope), vars)
+                Puppet.warning("#{compile_options}")
+                # We have to parse the AST inside the compiler block, because it
+                # initializes the necessary type loaders for us.
+                # See https://github.com/puppetlabs/bolt/pull/1748 for more details
+                ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+                # Node definitions must be at the top level of the apply block.
+                # That means the apply body either a) consists of just a
+                # NodeDefinition, b) consists of a BlockExpression which may
+                # contain NodeDefinitions, or c) doesn't contain NodeDefinitions.
+                # See https://github.com/puppetlabs/bolt/pull/1512 for more details
+                definitions = if ast.is_a?(Puppet::Pops::Model::BlockExpression)
+                                ast.statements.select { |st| st.is_a?(Puppet::Pops::Model::NodeDefinition) }
+                              elsif ast.is_a?(Puppet::Pops::Model::NodeDefinition)
+                                [ast]
+                              else
+                                []
+                              end
+                ast = Puppet::Pops::Model::Factory.PROGRAM(ast, definitions, ast.locator).model
```

Comment:
Would it be better to call this something else, instead of redefining `ast`?

Reply:
I dont think so. Its still AST. It just transformed.

Reply:
you could call it `transformed_ast`?

Reply:
Actually I do think this is an improvement. (Though I don't think variable reassignment/mutations in inherently bad). In bolt we need to conditionally do the transformation because AST snippets from apply blocks do not have a full tree with Program at the root. Where manifests applied with `bolt apply` _will_ have the full tree. (https://github.com/puppetlabs/bolt/commit/3a7597dda25cdb25854c7d08d37c5c58ab6a016b) Here we know we will always need to do it. I will update this and perhaps add a comment as to why.